File: test/config.test.ts
import { afterAll, beforeAll, beforeEach, describe, expect, test } from "bun:test";
import { existsSync, mkdirSync, readFileSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

const TEST_CONFIG_DIR = join(tmpdir(), `agent-mongo-config-test-${Date.now()}`);

// Set XDG before importing config module so it uses our temp dir
process.env.XDG_CONFIG_HOME = TEST_CONFIG_DIR;

const {
  readConfig,
  writeConfig,
  storeConnection,
  getConnection,
  getConnections,
  getDefaultConnectionAlias,
  removeConnection,
  setDefaultConnection,
  getSettings,
  getSetting,
  updateSetting,
  resetSettings,
} = await import("../src/lib/config.ts");

describe("config CRUD", () => {
  beforeAll(() => {
    process.env.XDG_CONFIG_HOME = TEST_CONFIG_DIR;
    mkdirSync(join(TEST_CONFIG_DIR, "agent-mongo"), { recursive: true });
    writeConfig({});
  });

  afterAll(() => {
    rmSync(TEST_CONFIG_DIR, { recursive: true, force: true });
    delete process.env.XDG_CONFIG_HOME;
  });

  test("readConfig returns empty object when no config exists", () => {
    rmSync(join(TEST_CONFIG_DIR, "agent-mongo", "config.json"), { force: true });
    expect(readConfig()).toEqual({});
  });

  test("writeConfig creates config file", () => {
    writeConfig({ default_connection: "test" });
    const configPath = join(TEST_CONFIG_DIR, "agent-mongo", "config.json");
    expect(existsSync(configPath)).toBe(true);
    const raw = JSON.parse(readFileSync(configPath, "utf8"));
    expect(raw.default_connection).toBe("test");
  });

  test("readConfig reads back what was written", () => {
    writeConfig({ default_connection: "prod", connections: { prod: { connection_string: "mongodb://localhost" } } });
    const config = readConfig();
    expect(config.default_connection).toBe("prod");
    expect(config.connections?.prod?.connection_string).toBe("mongodb://localhost");
  });
});

describe("connections", () => {
  beforeEach(() => {
    process.env.XDG_CONFIG_HOME = TEST_CONFIG_DIR;
    mkdirSync(join(TEST_CONFIG_DIR, "agent-mongo"), { recursive: true });
    writeConfig({});
  });

  afterAll(() => {
    rmSync(TEST_CONFIG_DIR, { recursive: true, force: true });
    delete process.env.XDG_CONFIG_HOME;
  });

  test("getConnections returns empty object when none stored", () => {
    expect(getConnections()).toEqual({});
  });

  test("storeConnection stores a connection and sets it as default", () => {
    storeConnection("local", { connection_string: "mongodb://localhost:27017" });
    expect(getConnection("local")).toEqual({ connection_string: "mongodb://localhost:27017" });
    expect(getDefaultConnectionAlias()).toBe("local");
  });

  test("storeConnection does not overwrite existing default", () => {
    storeConnection("local", { connection_string: "mongodb://localhost:27017" });
    storeConnection("prod", { connection_string: "mongodb://prod:27017", name: "Production" });
    expect(getDefaultConnectionAlias()).toBe("local");
    expect(Object.keys(getConnections())).toEqual(["local", "prod"]);
  });

  test("storeConnection with name and database", () => {
    storeConnection("dev", {
      connection_string: "mongodb://dev:27017",
      name: "Development",
      database: "mydb",
    });
    const conn = getConnection("dev");
    expect(conn?.name).toBe("Development");
    expect(conn?.database).toBe("mydb");
  });

  test("getConnection returns undefined for unknown alias", () => {
    expect(getConnection("nonexistent")).toBeUndefined();
  });

  test("setDefaultConnection switches the active connection", () => {
    storeConnection("a", { connection_string: "mongodb://a" });
    storeConnection("b", { connection_string: "mongodb://b" });
    expect(getDefaultConnectionAlias()).toBe("a");
    setDefaultConnection("b");
    expect(getDefaultConnectionAlias()).toBe("b");
  });

  test("setDefaultConnection throws for unknown alias", () => {
    expect(() => setDefaultConnection("nonexistent")).toThrow("Unknown connection");
  });

  test("removeConnection removes and reassigns default", () => {
    storeConnection("first", { connection_string: "mongodb://first" });
    storeConnection("second", { connection_string: "mongodb://second" });
    removeConnection("first");
    expect(getDefaultConnectionAlias()).toBe("second");
    expect(getConnection("first")).toBeUndefined();
  });

  test("removeConnection clears default when last connection removed", () => {
    storeConnection("only", { connection_string: "mongodb://only" });
    removeConnection("only");
    expect(getDefaultConnectionAlias()).toBeUndefined();
    expect(getConnections()).toEqual({});
  });

  test("removeConnection throws for unknown alias", () => {
    expect(() => removeConnection("nonexistent")).toThrow("Unknown connection");
  });
});

describe("settings", () => {
  beforeEach(() => {
    process.env.XDG_CONFIG_HOME = TEST_CONFIG_DIR;
    mkdirSync(join(TEST_CONFIG_DIR, "agent-mongo"), { recursive: true });
    writeConfig({});
  });

  afterAll(() => {
    rmSync(TEST_CONFIG_DIR, { recursive: true, force: true });
    delete process.env.XDG_CONFIG_HOME;
  });

  test("getSettings returns empty object when no settings stored", () => {
    expect(getSettings()).toEqual({});
  });

  test("updateSetting persists a top-level setting", () => {
    updateSetting("defaults.limit", 50);
    expect(getSetting("defaults.limit")).toBe(50);
  });

  test("updateSetting persists nested settings", () => {
    updateSetting("query.timeout", 5000);
    expect(getSetting("query.timeout")).toBe(5000);
  });

  test("updateSetting creates intermediate objects", () => {
    updateSetting("truncation.maxLength", 300);
    const settings = getSettings();
    expect(settings.truncation?.maxLength).toBe(300);
  });

  test("getSetting returns undefined for non-existent key", () => {
    expect(getSetting("nonexistent.key")).toBeUndefined();
  });

  test("getSetting traverses dotted paths", () => {
    updateSetting("defaults.limit", 25);
    updateSetting("defaults.sampleSize", 100);
    expect(getSetting("defaults.limit")).toBe(25);
    expect(getSetting("defaults.sampleSize")).toBe(100);
  });

  test("resetSettings clears all settings", () => {
    updateSetting("defaults.limit", 50);
    updateSetting("query.timeout", 5000);
    resetSettings();
    expect(getSettings()).toEqual({});
  });

  test("settings isolation: updateSetting does not touch connection data", () => {
    storeConnection("test", { connection_string: "mongodb://test" });
    updateSetting("defaults.limit", 10);
    expect(getConnection("test")).toEqual({ connection_string: "mongodb://test" });
    expect(getSetting("defaults.limit")).toBe(10);
  });
});

File: internal/config/config_test.go
package config

import (
	"strings"
	"testing"
)

// isolate points XDG_CONFIG_HOME at a fresh temp dir so each test reads and
// writes its own config.json. Read() reads the file fresh every call, so no
// cache clearing is needed.
func isolate(t *testing.T) {
	t.Helper()
	t.Setenv("XDG_CONFIG_HOME", t.TempDir())
}

func TestReadReturnsEmptyWhenNoConfig(t *testing.T) {
	isolate(t)
	cfg := Read()
	if len(cfg.Connections) != 0 {
		t.Errorf("expected 0 connections, got %d", len(cfg.Connections))
	}
	if cfg.DefaultConnection != "" {
		t.Errorf("default = %q, want empty", cfg.DefaultConnection)
	}
	if cfg.Settings != nil {
		t.Errorf("settings = %+v, want nil", cfg.Settings)
	}
}

func TestWriteCreatesConfigFile(t *testing.T) {
	isolate(t)
	if err := Write(Config{DefaultConnection: "test"}); err != nil {
		t.Fatalf("Write() error: %v", err)
	}
	if got := Read().DefaultConnection; got != "test" {
		t.Errorf("default = %q, want %q", got, "test")
	}
}

func TestReadRoundTripsWhatWasWritten(t *testing.T) {
	isolate(t)
	err := Write(Config{
		DefaultConnection: "prod",
		Connections: map[string]Connection{
			"prod": {ConnectionString: "mongodb://localhost"},
		},
	})
	if err != nil {
		t.Fatalf("Write() error: %v", err)
	}
	cfg := Read()
	if cfg.DefaultConnection != "prod" {
		t.Errorf("default = %q, want %q", cfg.DefaultConnection, "prod")
	}
	if got := cfg.Connections["prod"].ConnectionString; got != "mongodb://localhost" {
		t.Errorf("connection_string = %q, want %q", got, "mongodb://localhost")
	}
}

func TestConnectionsReturnsEmptyWhenNoneStored(t *testing.T) {
	isolate(t)
	if got := Connections(); len(got) != 0 {
		t.Errorf("Connections() = %v, want empty", got)
	}
}

func TestStoreConnectionStoresAndSetsDefault(t *testing.T) {
	isolate(t)
	if err := StoreConnection("local", Connection{ConnectionString: "mongodb://localhost:27017"}); err != nil {
		t.Fatalf("StoreConnection() error: %v", err)
	}
	conn, ok := GetConnection("local")
	if !ok {
		t.Fatal("GetConnection(local) not found")
	}
	if conn.ConnectionString != "mongodb://localhost:27017" {
		t.Errorf("connection_string = %q", conn.ConnectionString)
	}
	if got := DefaultConnectionAlias(); got != "local" {
		t.Errorf("default = %q, want %q", got, "local")
	}
}

func TestStoreConnectionDoesNotOverwriteExistingDefault(t *testing.T) {
	isolate(t)
	if err := StoreConnection("local", Connection{ConnectionString: "mongodb://localhost:27017"}); err != nil {
		t.Fatalf("StoreConnection(local) error: %v", err)
	}
	err := StoreConnection("prod", Connection{ConnectionString: "mongodb://prod:27017", Name: "Production"})
	if err != nil {
		t.Fatalf("StoreConnection(prod) error: %v", err)
	}
	if got := DefaultConnectionAlias(); got != "local" {
		t.Errorf("default = %q, want %q", got, "local")
	}
	if got := ConnectionAliases(); len(got) != 2 || got[0] != "local" || got[1] != "prod" {
		t.Errorf("aliases = %v, want [local prod]", got)
	}
}

func TestStoreConnectionWithNameAndDatabase(t *testing.T) {
	isolate(t)
	err := StoreConnection("dev", Connection{
		ConnectionString: "mongodb://dev:27017",
		Name:             "Development",
		Database:         "mydb",
	})
	if err != nil {
		t.Fatalf("StoreConnection() error: %v", err)
	}
	conn, ok := GetConnection("dev")
	if !ok {
		t.Fatal("GetConnection(dev) not found")
	}
	if conn.Name != "Development" {
		t.Errorf("name = %q, want %q", conn.Name, "Development")
	}
	if conn.Database != "mydb" {
		t.Errorf("database = %q, want %q", conn.Database, "mydb")
	}
}

func TestGetConnectionReturnsFalseForUnknownAlias(t *testing.T) {
	isolate(t)
	if _, ok := GetConnection("nonexistent"); ok {
		t.Error("GetConnection(nonexistent) = ok, want not found")
	}
}

func TestSetDefaultConnectionSwitchesActive(t *testing.T) {
	isolate(t)
	mustStore(t, "a", "mongodb://a")
	mustStore(t, "b", "mongodb://b")
	if got := DefaultConnectionAlias(); got != "a" {
		t.Fatalf("default = %q, want %q", got, "a")
	}
	if err := SetDefaultConnection("b"); err != nil {
		t.Fatalf("SetDefaultConnection() error: %v", err)
	}
	if got := DefaultConnectionAlias(); got != "b" {
		t.Errorf("default = %q, want %q", got, "b")
	}
}

func TestSetDefaultConnectionErrorsForUnknownAlias(t *testing.T) {
	isolate(t)
	err := SetDefaultConnection("nonexistent")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "Unknown connection") {
		t.Errorf("error = %q, want it to mention 'Unknown connection'", err)
	}
}

func TestRemoveConnectionRemovesAndReassignsDefault(t *testing.T) {
	isolate(t)
	mustStore(t, "first", "mongodb://first")
	mustStore(t, "second", "mongodb://second")
	if err := RemoveConnection("first"); err != nil {
		t.Fatalf("RemoveConnection() error: %v", err)
	}
	if got := DefaultConnectionAlias(); got != "second" {
		t.Errorf("default = %q, want %q", got, "second")
	}
	if _, ok := GetConnection("first"); ok {
		t.Error("GetConnection(first) = ok, want removed")
	}
}

func TestRemoveConnectionClearsDefaultWhenLastRemoved(t *testing.T) {
	isolate(t)
	mustStore(t, "only", "mongodb://only")
	if err := RemoveConnection("only"); err != nil {
		t.Fatalf("RemoveConnection() error: %v", err)
	}
	if got := DefaultConnectionAlias(); got != "" {
		t.Errorf("default = %q, want empty", got)
	}
	if got := Connections(); len(got) != 0 {
		t.Errorf("Connections() = %v, want empty", got)
	}
}

func TestRemoveConnectionErrorsForUnknownAlias(t *testing.T) {
	isolate(t)
	err := RemoveConnection("nonexistent")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "Unknown connection") {
		t.Errorf("error = %q, want it to mention 'Unknown connection'", err)
	}
}

func mustStore(t *testing.T, alias, uri string) {
	t.Helper()
	if err := StoreConnection(alias, Connection{ConnectionString: uri}); err != nil {
		t.Fatalf("StoreConnection(%s) error: %v", alias, err)
	}
}
